Add API for modifying module data

sergeuz opened the pull request:
This PR implements a basic API for modifying module data:

- `updateModuleSha256()` – computes and updates the module's SHA-256 hash.
- `updateModuleCrc32()` – computes and updates the module's CRC32 checksum.
- `updateModulePrefix()` – rewrites the module's prefix data.
- `updateModuleData()` – rewrites the firmware part of the module data (i.e. everything between the prefix and suffix blocks).
- `extractModuleData()` – extracts the firmware part of the module data.

sergeuz commented:
@monkbroc @busticated I'm re-requesting a review since the API has changed a little – we decided to implement high-level APIs for module compression/decompression directly in `binary-version-reader`, so that DS is completely decoupled from the module format specifics.

The following functions have been added since the last review:
- `compressModule()`
- `decompressModule()`
- `combineModules()`
- `splitCombinedModules()`

These two have been removed:
- `updateModuleData()`
- `extractModuleData()`

monkbroc commented on lib/moduleEncoding.js:
```diff
@@ -0,0 +1,558 @@
+		dataSize -= MODULE_PREFIX_SIZE;
+	}
+	// Compress the module data
+	// TODO: Investigate compressing the module data in place to optimize memory usage, or use some
```
Remove this comment. If you're not doing it now, I doubt anybody is going to come back and do this. It also feel like premature optimization compared to all the other memory allocations that are happening in device service. A more meaningful optimization would be to store the compressed binary back in the database for system firmware and product firmware such that the next device requesting a compressed binary will not trigger the compression operation in device service.

sergeuz replied:
Why? It's just a comment pointing out something that can be improved, not a request for action :slightly_smiling_face: While in-place compression is indeed a convoluted approach, using liked buffers makes sense in both scenarios, especially for combined modules. After all, we just need to send this data to the client, it doesn't matter whether the buffer is contiguous.

sergeuz commented:
> given the nature of the code, i would love to see >90% branch coverage

@busticated Those untested branches are almost exclusively various checks for missing properties, which are quite pedantic in the new code. The actual logic of the encoding functions is fully covered by using test binaries bundled with the tests.

sergeuz commented:
Changes since the last review:

- Added separate methods to parse a module's prefix (`parsePrefix()`) and suffix data (`parseSuffix()`). These are useful when you need to retrieve some info from a module, e.g. the version number, without performing any integrity checks.
- The function used for CRC-32 computations can now be provided externally (we'll use a native implementation in Device Service).
- Extended product IDs are now parsed/encoded as a single 32-bit little-endian integer.
- Improved test coverage.

@monkbroc @busticated No re-review is required, as the changes are pretty minimal and well-tested both here and in the Device Service. I'm going to merge this and publish v1.0.0.
